README
Welcome to the BeeCrypt crypto library!

Copyright (c) 1997-2000 Virtual Unlimited B.V.

Author: Bob Deblier <bob@virtualunlimited.com>

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2.1 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

For the specifics of this license, see file 'LICENSE', included in this
distribution.



About BeeCrypt:

This library is an ongoing project of Virtual Unlimited B.V. Its goal is to
provide strong and fast cryptography for use by our products, but we're
not limiting the use to that. We're releasing it under the LGPL license,
because we feel that cryptography should be open to inspection by everybody,
and available for use by everybody to safeguard privacy.

Note that depending on where you are, the use of cryptography may be limited
or forbidden by law. Before using this library, make sure you are legally
entitled to do so.

For more on Virtual Unlimited B.V. and our products please consult our website:

http://www.virtualunlimited.com

<plug>
Most of the algorithms are implemented from reliable sources such as:

"Handbook of Applied Cryptography"
 Alfred J. Menezes, Paul C. van Oorschot, Scott A. Vanstone
 CRC Press

"Applied Cryptography", second edition
 Bruce Schneier
 Wiley

For crypto enthusiasts these books are invaluable background material.

IEEE P1363 "Standard Specifications for Public Key Cryptography" is a very
interesting draft standard, which we will try to comply with.
</plug>

The structures in the library are geared towards exchange with Java
and its security and cryptography classes. We already have quite a number
of cryptographic classes developed in Java, which will be released in the
future in another open source project.

Included in the library are:
    - entropy sources for initializing pseudo-random generators
    - pseudo-random generators: FIPS-186, Mersenne Twister
    - block ciphers: Blowfish
    - hash functions: SHA-1
    - keyed hash functions: SHA-1/HMAC
    - multi-precision integer library, with assembler-optimized routines
    - probabilistic primality testing, with optimized small prime trial division
    - discrete logarithm parameter generation over a prime field
    - Diffie-Hellman key agreement
    - ElGamal signature scheme (two variants)

Planned for the near future are:
    - compliance with and compliance statements for IEEE P1363
	- DHAES
    - DSA (i.e. the rest of FIPS-186)
    - more blockciphers (Twofish, Rijndael, Misty1, ... )
	- more hash functions (RIPEMD-160, HAVAL, ... )
	- Elliptic Curves (ECDSA, ... )

The library has been tested on the following platforms:
    - Solaris 2.6/2.7 Sparc (with Workshop compilers, gnu binutils and make)
        - includes assembler optimization for ultrasparc
	- Solaris 2.7 Intel (with gnu compiler, binutils and make)
		- includes assembler optimization (tuned for Pentium Pro and higher)
    - Linux glibc 2.x x86
        - includes assembler optimization (tuned for Pentium Pro and higher)
    - Linux glibc 2.x ppc
        - includes assembler optimization (for generic 32-bit PowerPC)
	- FreeBSD 4.0 x86
		- includes assembler optimization (tuned for Pentium Pro and higher)
    - Win32 (Windows 95, 98, NT 4.0, 2000)
        - includes assembler optimization (tuned for Pentium Pro and higher)

The library is currently in the process of being ported to:
	- Cygwin
	- Darwin 1.0

Consult file 'DOCUMENTATION' on how to use the library. There is also a program
called 'beetest.c' included which gives an example of how to use the API. You
can also use it to test whether your compiled library works.

If you want to report bugs, make suggestions, contribute fixes or
enhancements, please see our website: http://beecrypt.virtualunlimited.com
or contact me at beecrypt@virtualunlimited.com

Sincerely

Bob Deblier


NEWS
4.1.1:
	- Fixed shared library version info.

4.1.0:
	- Added SHA-384 and SHA-512 algorithms.
	- Added HMAC-SHA-384 and HMAC-SHA-512 algorithms.
	- Added generic SSE2 optimization for the above algorithms.
	- Added more digest algorithms for PKCS#1 EMSA.
	- Optimized swap32 and swap64 routines on Linux.
	- Fixed missing definition in mpopt.h for s390x.
	- Fixed nostackexec configuration bug.
	- Fixed problem in Date::toString.
	- Fixed deadlock problem which occured in certain cases where security
	  or crypto SPI constructor called getInstance for another security or
	  crypto SPI.
	- Fixed a bug in the generic CBC encryption code; when called with
	  nblocks == 1, the feedback was set incorrectly.
	- Fixed a bug in mpbsubmod; sometimes it takes multiple additions of
	  the modulus to get a positive number.
	- Fixed PowerPC 64-bit configuration problem on Linux.

4.0.0:
	- Added a C++ API interface, modeled after Java's security & crypto API.
	- Added the new GNU noexecstack feature.
	- Added more x86_64 and s390x assembler routines.
	- Modified i2osp, so that it only requires as many octets as there are
	  significant bytes in the multi-precision integers.
	- Fixed a bug in the creation of rsa keypairs; code was not correctly
	  migrated to new calling sequence. The code now implements the method
	  described in IEEE P.1363.
	- Fixed another bug in mpextgcd_w which sometimes returned incorrect
	  results.
	- Fixed a bug in mprshiftlsz, which didn't work correctly when size = 1.
	- Fixed a configuration problem on Tru64 Unix.

3.1.0:
	- Added wiping of private key components of keypairs before freeing.
	- Fixed bug in mpextgcd_w which sometimes returned incorrect result.
	- Fixed error in PowerPC 64-bit assembler symbol definitions.

3.0.0:
	- Cleaned up installed header files.
	- Modified the API so that all keys can be passed as arrays of bytes.
	- Modified the API so that all key sizes are given in bits.
	- Modified the multi-precision integer library to work better on 64-bit
	  machines.
	- Modified the assembly source generation mechanism, employing the m4
	  macro processor.
	- Added multi-precision integer vectorized assembler routines for
	  Itanium.
	- Added multi-precision integer assembler routines for PowerPC 64-bit.
	- Added multi-precision integer assembler routines for Alpha.
	- Added multi-precision integer assembler routines for Opteron.
	- Added multi-precision integer assembler routines for IBM zSeries 64-bit.
	- Added multi-precision integer assembler routines for M68K.
	- Added Jeff Johnson's python bindings.
	- Added new unit tests.
	- Added new benchmarking programs.

2.3.0pre:
	- Modified the header files so that the library now uses self-contained autoconf-generated configuration files; a program employing BeeCrypt can now use the symbols already tested and defined instead of having to regenerate them (thus also eliminating the risk of inconsistencies).
	- Added the AES algorithm, with assembler routines for i586 and powerpc.
	- Added the DSA signature algorithm.
	- Added PowerPC assembler routines for blowfish.
	- Added Pentium4 SSE2 assembler multiplication routines.
	- Fixed the RSA CRT algorithm.
	- Fixed the gas/i386 mp32even and mp32odd routines.
	- Fixed a bug in modular inverse computation; thanks to Jeff Johnson of RedHat for pointing this out.
	- Fixed a bug in testing the result of a gcd operation in the mp32prndconone routine.
	- Fixed an ugly bug in base64 decoding.
	- Fixed compatibility with the latest automake & autoconf versions.
	- Replaces CPU optimization mechanism in configure script.

2.1.0:
	- Added support for automake, autoheader and libtool, which should make compiling the library even easier.
	- Changed DHAES API to conform to IEEE P.1363 submission and to allow for uneven key splitting.
	- Improved PKCS#5 padding routines.
	- Added a hash reset to the hashFunctionContextInit function. This was pointed out by Marko Kreen.
	- Fixed problem with configuring on i486-pc-linux-gnu. This was pointed out Steve O'Neill.
	- Fixed problem in the C version of mp32sub where carry would sometimes be missed. This was pointed out by Jon Sturgeon.

mp.h
/*
 * Copyright (c) 2002, 2003 Bob Deblier
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

/*!\file mp.h
 * \brief Multi-precision integer routines.
 *
 * The routines declared here are all low-level operations, most of them
 * suitable to be implemented in assembler. Prime candidates are in order
 * of importance (according to gprof):
 * <ul>
 *  <li>mpaddmul
 *  <li>mpsetmul 
 *  <li>mpaddsqrtrc
 *  <li>mpsub
 *  <li>mpadd
 * </ul>
 *
 * With some smart use of available assembler instructions, it's possible
 * to speed these routines up by a factor of 2 to 4.
 *
 * \author Bob Deblier <bob.deblier@pandora.be>
 * \ingroup MP_m
 */

#ifndef _MP_H
#define _MP_H

#include "beecrypt.api.h"
#include "mpopt.h"

#define MP_HWBITS	(MP_WBITS >> 1)
#define MP_WBYTES	(MP_WBITS >> 3)
#define MP_WNIBBLES	(MP_WBITS >> 2)

#if (MP_WBITS == 64)
# define MP_WORDS_TO_BITS(x)	((x) << 6)
# define MP_WORDS_TO_NIBBLES(x)	((x) << 4)
# define MP_WORDS_TO_BYTES(x)	((x) << 3)
# define MP_BITS_TO_WORDS(x)	((x) >> 6)
# define MP_NIBBLES_TO_WORDS(x)	((x) >> 4)
# define MP_BYTES_TO_WORDS(x)	((x) >> 3)
#elif (MP_WBITS == 32)
# define MP_WORDS_TO_BITS(x)	((x) << 5)
# define MP_WORDS_TO_NIBBLES(x)	((x) << 3)
# define MP_WORDS_TO_BYTES(x)	((x) << 2)
# define MP_BITS_TO_WORDS(x)	((x) >> 5) 
# define MP_NIBBLES_TO_WORDS(x)	((x) >> 3)
# define MP_BYTES_TO_WORDS(x)	((x) >> 2)
#else
# error
#endif

#define MP_MSBMASK	(((mpw) 0x1) << (MP_WBITS-1))
#define MP_LSBMASK	 ((mpw) 0x1)
#define MP_ALLMASK	~((mpw) 0x0)

#ifdef __cplusplus
extern "C" {
#endif

#ifndef ASM_MPCOPY
# define mpcopy(size, dst, src) memcpy(dst, src, MP_WORDS_TO_BYTES(size))
#else
BEECRYPTAPI
void mpcopy(size_t size, mpw* dest, const mpw* src);
#endif

#ifndef ASM_MPMOVE
# define mpmove(size, dst, src) memmove(dst, src, MP_WORDS_TO_BYTES(size))
#else
BEECRYPTAPI
void mpmove(size_t size, mpw* dest, const mpw* src);
#endif

/*!\fn void mpzero(size_t size, mpw* data)
 * \brief This function zeroes a multi-precision integer of a given size.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 */
BEECRYPTAPI
void mpzero(size_t size, mpw* data);

/*!\fn void mpfill(size_t size, mpw* data, mpw fill)
 * \brief This function fills each word of a multi-precision integer with a
 *  given value.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \param fill The value fill the data with.
 */
BEECRYPTAPI
void mpfill(size_t size, mpw* data, mpw fill);

/*!\fn int mpodd(size_t size, const mpw* data)
 * \brief This functions tests if a multi-precision integer is odd.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if odd
 * \retval 0 if even
 */
BEECRYPTAPI
int mpodd (size_t size, const mpw* data);

/*!\fn int mpeven(size_t size, const mpw* data)
 * \brief This function tests if a multi-precision integer is even.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if even
 * \retval 0 if odd
 */
BEECRYPTAPI
int mpeven(size_t size, const mpw* data);

/*!\fn int mpz(size_t size, const mpw* data)
 * \brief This function tests if a multi-precision integer is zero.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if zero
 * \retval 0 if not zero
 */
BEECRYPTAPI
int mpz  (size_t size, const mpw* data);

/*!\fn int mpnz(size_t size, const mpw* data)
 * \brief This function tests if a multi-precision integer is not zero.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if not zero
 * \retval 0 if zero
 */
BEECRYPTAPI
int mpnz (size_t size, const mpw* data);

/*!\fn int mpeq(size_t size, const mpw* xdata, const mpw* ydata)
 * \brief This function tests if two multi-precision integers of the same size
 *  are equal.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if equal
 * \retval 0 if not equal
 */
BEECRYPTAPI
int mpeq (size_t size, const mpw* xdata, const mpw* ydata);

/*!\fn int mpne(size_t size, const mpw* xdata, const mpw* ydata)
 * \brief This function tests if two multi-precision integers of the same size
 *  differ.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if not equal
 * \retval 0 if equal
 */
BEECRYPTAPI
int mpne (size_t size, const mpw* xdata, const mpw* ydata);

/*!\fn int mpgt(size_t size, const mpw* xdata, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of the same size is greater than the second.
 * \note The comparison treats the arguments as unsigned.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if greater
 * \retval 0 if less or equal
 */
BEECRYPTAPI
int mpgt (size_t size, const mpw* xdata, const mpw* ydata);

/*!\fn int mplt(size_t size, const mpw* xdata, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of the same size is less than the second.
 * \note The comparison treats the arguments as unsigned.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if less
 * \retval 0 if greater or equal
 */
BEECRYPTAPI
int mplt (size_t size, const mpw* xdata, const mpw* ydata);

/*!\fn int mpge(size_t size, const mpw* xdata, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of the same size is greater than or equal to the second.
 * \note The comparison treats the arguments as unsigned.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if greater or equal
 * \retval 0 if less
 */
BEECRYPTAPI
int mpge (size_t size, const mpw* xdata, const mpw* ydata);

/*!\fn int mple(size_t size, const mpw* xdata, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of the same size is less than or equal to the second.
 * \note The comparison treats the arguments as unsigned.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if less or equal
 * \retval 0 if greater
 */
BEECRYPTAPI
int mple (size_t size, const mpw* xdata, const mpw* ydata);

/*!\fn int mpeqx(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function tests if two multi-precision integers of different
 *  size are equal.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if equal
 * \retval 0 if not equal
 */
BEECRYPTAPI
int mpeqx(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mpnex(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function tests if two multi-precision integers of different
 *  size are equal.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if equal
 * \retval 0 if not equal
*/
BEECRYPTAPI
int mpnex(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mpgtx(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of different size is greater than the second.
 * \note The comparison treats the arguments as unsigned.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the second multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if greater
 * \retval 0 if less or equal
 */
BEECRYPTAPI
int mpgtx(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mpltx(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of different size is less than the second.
 * \note The comparison treats the arguments as unsigned.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the second multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if less
 * \retval 0 if greater or equal
 */
BEECRYPTAPI
int mpltx(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mpgex(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of different size is greater than or equal to the second.
 * \note The comparison treats the arguments as unsigned.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the second multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if greater or equal
 * \retval 0 if less
 */
BEECRYPTAPI
int mpgex(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mplex(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function tests if the first of two multi-precision integers
 *  of different size is less than or equal to the second.
 * \note The comparison treats the arguments as unsigned.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the second multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if less or equal
 * \retval 0 if greater
 */
BEECRYPTAPI
int mplex(size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mpisone(size_t size, const mpw* data)
 * \brief This functions tests if the value of a multi-precision integer is
 *  equal to one.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if one
 * \retval 0 if not one
 */
BEECRYPTAPI
int mpisone(size_t size, const mpw* data);

/*!\fn int mpistwo(size_t size, const mpw* data)
 * \brief This function tests if the value of a multi-precision integer is
 *  equal to two.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if two
 * \retval 0 if not two
 */
BEECRYPTAPI
int mpistwo(size_t size, const mpw* data);

/*!\fn int mpleone(size_t size, const mpw* data);
 * \brief This function tests if the value of a multi-precision integer is
 *  less than or equal to one.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if less than or equal to one.
 * \retval 0 if greater than one.
 */
BEECRYPTAPI
int mpleone(size_t size, const mpw* data);

/*!\fn int mpeqmone(size_t size, const mpw* xdata, const mpw* ydata);
 * \brief This function tests if multi-precision integer x is equal to y
 *  minus one.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \retval 1 if less than or equal to one.
 * \retval 0 if greater than one.
 */
BEECRYPTAPI
int mpeqmone(size_t size, const mpw* xdata, const mpw* ydata);

/*!\fn int mpmsbset(size_t size, const mpw* data)
 * \brief This function tests if the most significant bit of a multi-precision
 *  integer is set.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if set
 * \retval 0 if not set
 */
BEECRYPTAPI
int mpmsbset(size_t size, const mpw* data);

/*!\fn int mplsbset(size_t size, const mpw* data)
 * \brief This function tests if the leiast significant bit of a multi-precision
 *  integer is set.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 * \retval 1 if set
 * \retval 0 if not set
 */
BEECRYPTAPI
int mplsbset(size_t size, const mpw* data);

/*!\fn void mpsetmsb(size_t size, mpw* data)
 * \brief This function sets the most significant bit of a multi-precision
 *  integer.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 */
BEECRYPTAPI
void mpsetmsb(size_t size, mpw* data);

/*!\fn void mpsetlsb(size_t size, mpw* data)
 * \brief This function sets the least significant bit of a multi-precision
 *  integer.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 */
BEECRYPTAPI
void mpsetlsb(size_t size, mpw* data);

/*!\fn void mpclrmsb(size_t size, mpw* data)
 * \brief This function clears the most significant bit of a multi-precision
 *  integer.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 */
BEECRYPTAPI
void mpclrmsb(size_t size, mpw* data);

/*!\fn void mpclrlsb(size_t size, mpw* data)
 * \brief This function clears the least significant bit of a multi-precision
 *  integer.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 */
BEECRYPTAPI
void mpclrlsb(size_t size, mpw* data);

BEECRYPTAPI
void mpand(size_t, mpw*, const mpw*);

BEECRYPTAPI
void mpor(size_t, mpw*, const mpw*);

BEECRYPTAPI
void mpxor(size_t, mpw*, const mpw*);

/*!\fn mpnot(size_t size, mpw* data)
 * \brief This function flips all bits of a multi-precision integer.
 * \param size The size of the multi-precision integer.
 * \param data The multi-precision integer data.
 */
BEECRYPTAPI
void mpnot(size_t size, mpw* data);

/*!\fn void mpsetw(size_t size, mpw* xdata, mpw y)
 * \brief This function sets the value of a multi-precision integer to the
 *  given word. The given value is copied into the least significant word,
 *  while the most significant words are zeroed.
 * \param size The size of the multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param y The multi-precision word.
 */
BEECRYPTAPI
void mpsetw(size_t size, mpw* xdata, mpw y);

/*!\fn void mpsetx(size_t xsize, mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function set the value of the first multi-precision integer
 *  to the second, truncating the most significant words if ysize > xsize, or
 *  zeroing the most significant words if ysize < xsize.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the second multi-precision integer.
 * \param ydata The second multi-precision integer.
 */
void mpsetx(size_t xsize, mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mpaddw(size_t size, mpw* xdata, mpw y)
 * \brief This function adds one word to a multi-precision integer.
 *  The performed operation is in pseudocode: x += y.
 * \param size The size of the multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param y The multi-precision word.
 * \return The carry-over value of the operation; this value is either 0 or 1.
 */
BEECRYPTAPI
int mpaddw(size_t size, mpw* xdata, mpw y);

/*!\fn int mpadd(size_t size, mpw* xdata, const mpw* ydata)
 * \brief This function adds two multi-precision integers of equal size.
 *  The performed operation is in pseudocode: x += y.
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \return The carry-over value of the operation; this value is either 0 or 1.
 */
BEECRYPTAPI
int mpadd (size_t size, mpw* xdata, const mpw* ydata);

/*!\fn int mpaddx(size_t xsize, mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function adds two multi-precision integers of different size.
 *  The performed operation in pseudocode: x += y.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the second multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \return The carry-over value of the operation; this value is either 0 or 1.
 */
BEECRYPTAPI
int mpaddx(size_t xsize, mpw* xdata, size_t ysize, const mpw* ydata);

/*!\fn int mpsubw(size_t size, mpw* xdata, mpw y)
 * \brief This function subtracts one word to a multi-precision integer.
 *  The performed operation in pseudocode: x -= y
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param y The multi-precision word.
 * \return The carry-over value of the operation; this value is either 0 or 1.
 */
BEECRYPTAPI
int mpsubw(size_t size, mpw* xdata, mpw y);

/*!\fn int mpsub(size_t size, mpw* xdata, const mpw* ydata)
 * \brief This function subtracts two multi-precision integers of equal size.
 *  The performed operation in pseudocode: x -= y
 * \param size The size of the multi-precision integers.
 * \param xdata The first multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \return The carry-over value of the operation; this value is either 0 or 1.
 */
BEECRYPTAPI
int mpsub (size_t size, mpw* xdata, const mpw* ydata);

/*!\fn int mpaddx(size_t xsize, mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function subtracts two multi-precision integers of different
 *  size. The performed operation in pseudocode: x -= y.
 * \param xsize The size of the first multi-precision integer.
 * \param xdata The first multi-precision integer.
 * \param ysize The size of the second multi-precision integer.
 * \param ydata The second multi-precision integer.
 * \return The carry-over value of the operation; this value is either 0 or 1.
 */
BEECRYPTAPI
int mpsubx(size_t xsize, mpw* xdata, size_t ysize, const mpw* ydata);

BEECRYPTAPI
int mpmultwo(size_t size, mpw* data);

BEECRYPTAPI
void mpneg(size_t size, mpw* data);

BEECRYPTAPI
size_t mpsize(size_t size, const mpw* data);

BEECRYPTAPI
size_t mpbits(size_t size, const mpw* data);

BEECRYPTAPI
size_t mpmszcnt(size_t size, const mpw* data);

BEECRYPTAPI
size_t mplszcnt(size_t size, const mpw* data);

BEECRYPTAPI
void mplshift(size_t size, mpw* data, size_t count);

BEECRYPTAPI
void mprshift(size_t size, mpw* data, size_t count);

BEECRYPTAPI
size_t mprshiftlsz(size_t size, mpw* data);

BEECRYPTAPI
size_t mpnorm(size_t size, mpw* data);

BEECRYPTAPI
void mpdivtwo (size_t size, mpw* data);

BEECRYPTAPI
void mpsdivtwo(size_t size, mpw* data);

/*!\fn mpw mpsetmul(size_t size, mpw* result, const mpw* data, mpw y)
 * \brief This function performs a multi-precision multiply-setup.
 *
 * This function is used in the computation of a full multi-precision
 * multiplication. By using it we can shave off a few cycles; otherwise we'd
 * have to zero the least significant half of the result first and use
 * another call to the slightly slower mpaddmul function.
 *
 * \param size The size of multi-precision integer multiplier.
 * \param result The place where result will be accumulated.
 * \param data The multi-precision integer multiplier.
 * \param y The multiplicand.
 * \return The carry-over multi-precision word.
 */
BEECRYPTAPI
mpw mpsetmul   (size_t size, mpw* result, const mpw* data, mpw y);

/*!\fn mpw mpaddmul(size_t size, mpw* result, const mpw* data, mpw y)
 * \brief This function performs a mult-precision multiply-accumulate.
 *
 * This function is used in the computation of a full multi-precision
 * multiplication. It computes the product-by-one-word and accumulates it with
 * the previous result.
 *
 * \param size The size of multi-precision integer multiplier.
 * \param result The place where result will be accumulated.
 * \param data The multi-precision integer multiplier.
 * \param y The multiplicand.
 * \retval The carry-over multi-precision word.
 */
BEECRYPTAPI
mpw mpaddmul   (size_t size, mpw* result, const mpw* data, mpw y);

/*!\fn void mpaddsqrtrc(size_t size, mpw* result, const mpw* data)
 * \brief This function is used in the calculation of a multi-precision
 * squaring.
 */
BEECRYPTAPI
void mpaddsqrtrc(size_t, mpw*, const mpw*);

/*!\fn void mpmul(mpw* result, size_t xsize, const mpw* xdata, size_t ysize, const mpw* ydata)
 * \brief This function computes a full multi-precision product.
 */
BEECRYPTAPI
void mpmul(mpw*, size_t, const mpw*, size_t, const mpw*);
/*!\fn void mpsqr(mpw* result, size_t size, const mpw* data)
 * \brief This function computes a full multi-precision square.
 */
BEECRYPTAPI
void mpsqr(mpw*, size_t, const mpw*);

BEECRYPTAPI
void mpgcd_w(size_t, const mpw*, const mpw*, mpw*, mpw*);
BEECRYPTAPI
int  mpextgcd_w(size_t, const mpw*, const mpw*, mpw*, mpw*);

BEECRYPTAPI
mpw mppndiv(mpw, mpw, mpw);

BEECRYPTAPI
void mpnmod(mpw*, size_t, const mpw*, size_t, const mpw*, mpw*);

BEECRYPTAPI
void mpmod (mpw*, size_t, const mpw*, size_t, const mpw*, mpw*);

BEECRYPTAPI
void mpndivmod(mpw*, size_t, const mpw*, size_t, const mpw*, mpw*);

/*
 * Output Routines
 */

BEECRYPTAPI
void mpprint(size_t size, const mpw* data);

BEECRYPTAPI
void mpprintln(size_t size, const mpw* data);

BEECRYPTAPI
void mpfprint(FILE* f, size_t size, const mpw* data);

BEECRYPTAPI
void mpfprintln(FILE* f, size_t size, const mpw* data);

/*
 * Conversion Routines
 */

BEECRYPTAPI
int os2ip(mpw*, size_t, const byte*, size_t);

BEECRYPTAPI
int i2osp(byte*, size_t, const mpw*, size_t);

BEECRYPTAPI
int hs2ip(mpw*, size_t, const char*, size_t);

#ifdef __cplusplus
}
#endif

#endif
